### DISCUSSION TITLE
Seems to be a bug with returning clause and scalars

### DISCUSSION THREAD

**Discussion by gdoumenc** [sqlalchemy/sqlalchemy#11896] [Usage Questions] [answered]
Hi,

I have an issue this a class using mixing : it seems that when doing an update with a returning clause then the scalars function fails with a `NoSuchColumnError` error.

If `cls` is a class with a mixin, then this code fails (If the class is not using mixin it works) : 

```
stmt = update(cls).filter(*criterions).values(**values).returning(cls)
query = db.session.scalars(stmt).unique()

```
But this one works : 

```
stmt = update(cls).filter(*criterions).values(**values)
query = db.session.execute(stmt)
```

Perhaps it's because I missed something

**Comment by CaselIT** [comment]
Hi,

Can you provide a runnable example that shows what is cls?

**Comment by gdoumenc** [comment]
Oups sorry,

Simplified version :

```
class DatedFrozenMixin:
    id: Mapped[int] = mapped_column(primary_key=True)
    created: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    modified: Mapped[datetime] = mapped_column(default=datetime.utcnow, onupdate=datetime.utcnow)


class AddressSQL(db.Model, DatedFrozenMixin)
    first_name: Mapped[str]
    last_name: Mapped[str]
    company: Mapped[str | None]
```

**Comment by CaselIT** [comment]
Can't really reproduce. Can you change the following runnable example to reproduce your problem?
```py
from datetime import datetime
import sqlalchemy as sa
from sqlalchemy.orm import Mapped, mapped_column, DeclarativeBase, Session


class Base(DeclarativeBase):
    pass


class DatedFrozenMixin:
    id: Mapped[int] = mapped_column(primary_key=True)
    created: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    modified: Mapped[datetime] = mapped_column(default=datetime.utcnow, onupdate=datetime.utcnow)


class AddressSQL(Base, DatedFrozenMixin):
    __tablename__ = "address"
    first_name: Mapped[str]
    last_name: Mapped[str]
    company: Mapped[str | None]


engine = sa.create_engine("sqlite:///:memory:")

Base.metadata.create_all(engine)
with engine.begin() as c:
    c.execute(sa.insert(AddressSQL).values(first_name="John", last_name="Doe", company="ACME"))

with Session(engine) as s, s.begin():
    stmt = (
        sa.update(AddressSQL)
        .filter(AddressSQL.first_name == "John")
        .values(company="boo")
        .returning(AddressSQL)
    )
    query = s.scalars(stmt).unique()
    print(query.all())
```

**Comment by gdoumenc** [comment]
Sure I will spend time to help understanding the issue... As I understand that I should not have the problem.

**Comment by gdoumenc** [comment]
I found something (I use my code as your example works..) :

```py
import os
import ssl
from datetime import datetime

from dotenv import load_dotenv
from sqlalchemy import create_engine
from sqlalchemy import update
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import Session
from sqlalchemy.orm import mapped_column

load_dotenv(".env.test")


class Base(DeclarativeBase):
    pass


class DatedFrozenMixin:
    id: Mapped[int] = mapped_column(primary_key=True)
    created: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    modified: Mapped[datetime] = mapped_column(default=datetime.utcnow, onupdate=datetime.utcnow)
    email: Mapped[str]
    is_removed: Mapped[bool] = mapped_column(default=False)
    frozen: Mapped[bool] = mapped_column(default=False)


class AddressSQL(Base, DatedFrozenMixin):
    __tablename__ = 'address_finalcustomeraddress'

    first_name: Mapped[str]
    last_name: Mapped[str]
    company: Mapped[str | None]
    address_line1: Mapped[str]
    address_line2: Mapped[str | None]
    phone: Mapped[str]
    postal_code: Mapped[str]
    city: Mapped[str]
    country: Mapped[str] = mapped_column(default="France")
    email: Mapped[str]
    designation: Mapped[str] = mapped_column(comment="Address name")
    instructions: Mapped[str]
    vat: Mapped[float] = mapped_column(default=0.2)


class BillingAddressSQL(Base, DatedFrozenMixin):
    __tablename__ = 'address_finalcustomeraddress'
    __table_args__ = {'extend_existing': True}


if __name__ == '__main__':
    # SSL context with certificate associated
    prod_ssl_context = ssl.create_default_context()
    prod_ssl_context.verify_mode = ssl.CERT_REQUIRED
    prod_ssl_loc = "./certificates/prod.pem"
    prod_ssl_context.load_verify_locations(prod_ssl_loc)

    uri = os.getenv('PROD_NEOREZO_URI')
    engine = create_engine(uri, connect_args={'ssl_context': prod_ssl_context},
                           execution_options={'schema_translate_map': {None: "tenant_parebrise123"}})

    stmt = update(AddressSQL).filter(AddressSQL.id == 68).values(first_name="123").returning(AddressSQL)
    with Session(engine) as s, s.begin():
        query = s.scalars(stmt).unique()
        print(query.all())
```

I have the error because of defining subclasses, here `BillingAddressSQL`. If I remove the subclass it works.

What is wrong with my code ?

**Reply by CaselIT** [reply]
I'm not sure this is something supported. @zzzeek can you also take a look? The strange bit is re-using the same table name with `BillingAddressSQL` but with less cols

I don't follow why you are trying to do. Could you explain the use case?

**Reply by zzzeek** [reply]
the mapping here is not supported and by redefining "address_finalcustomeraddress" the primary key column of the table to which AddressSQL is mapped is getting lost.     

If you want to re-use a table for more than one mapping use `__table__`:

```py
class BillingAddressSQL(Base, DatedFrozenMixin):
    __table__ = AddressSQL.__table__
```

**Comment by gdoumenc** [comment]
Thanks,

I've done that because I have 2 kinds of addresses : billing and shipping and for both the constraints are not same so I've put them on the model and not the database table : for example phone may be null on billing but not on shipping, and both have foreign keys to other elements.

So I understand that I've made a wrong usage of `extend_existing`, using `__table__ = AddressSQL.__table__` is working like a charm.

Do you think it is not a good idea to have subclasses for a table to add specific constraints on specific usage? I think this may be a common pattern

Any way thanks you very much for your answers

**Reply by CaselIT** [reply, marked as answer]
Contraints here refer to what exactly? Python checks or sql ones?

The issue witg your approach is that there is no way for sqlalchemy to discriminate between the two types, so a select all of BillingAddressSQL will return every row of the table.

I think that in this case a better solution may be:
- use two different table in the db. Likely the easiest solution.
- use inheritance in the orm. Have BillingAddressSQL be the base class and AddressSQL the derived class. Using same table inheritance you also have a very similar db table layout to the current one. You may also map DatedFrozenMixin as the base and the two addresses as derived classes. Docs are here https://docs.sqlalchemy.org/en/20/orm/inheritance.html#single-table-inheritance

**Comment by gdoumenc** [comment]
Constraints refer to python check.

Of course two different tables should be better but ... life is never like we expected.

Many many thanks for the link, I don't understand how I haven't see it before...🙏
